File: app/src/main/java/com/summerland/android/game/MainThread.java
package com.summerland.android.game;

import android.graphics.Canvas;
import android.view.SurfaceHolder;

/**
 * MainThread class, Created by steve on 30/09/16.
 */
public class MainThread extends Thread {

    private static final int FPS = 30;
    //private double averageFPS;
    private final SurfaceHolder surfaceHolder;
    private GamePanel gamePanel;
    private boolean running;

    MainThread(SurfaceHolder surfaceHolder, GamePanel gamePanel) {
        super();
        this.surfaceHolder = surfaceHolder;
        this.gamePanel = gamePanel;
    }

    @Override
    public void run(){
        long startTime, timeMillis, waitTime;
        //long totalTime = 0, frameCount = 0;
        long targetTime = 1000/FPS;

        while (running) {
            startTime = System.nanoTime();
            Canvas canvas = null;

            // try locking the canvas for pixel editing
            try {
                canvas = this.surfaceHolder.lockCanvas();
                synchronized (surfaceHolder) {
                    this.gamePanel.update();
                    this.gamePanel.draw(canvas);

                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                if (canvas != null)
                    try {
                        surfaceHolder.unlockCanvasAndPost(canvas);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
            }

            timeMillis = (System.nanoTime() - startTime)/1000000;
            waitTime = targetTime - timeMillis;

            if (waitTime >= 0) {
                try {
                    MainThread.sleep(waitTime);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            //totalTime += System.nanoTime() - startTime;
            //frameCount++;
            //if (frameCount == FPS) {
            //    averageFPS = 1000/((totalTime/frameCount)/1000000);
            //    totalTime = 0;
            //    frameCount = 0;
            //    System.out.println("Average FPS: " + averageFPS);
            //}
        }
    }

    void setRunning(boolean run) {
        running = run;
    }

    protected Object clone() throws CloneNotSupportedException {
        return super.clone();
    }
}

File: app/src/main/java/com/summerland/android/game/MainThread.kt
package com.summerland.android.game

import android.graphics.Canvas
import android.view.SurfaceHolder

/**
 * MainThread class, Created by steve on 30/09/16.
 */
class MainThread internal constructor(private val surfaceHolder: SurfaceHolder, private val gamePanel: GamePanel) : Thread() {

    override fun run() {
        var startTime: Long
        var timeMillis: Long
        var waitTime: Long
        //long totalTime = 0, frameCount = 0;
        val targetTime = (1000 / FPS).toLong()

        while (running) {
            startTime = System.nanoTime()
            var canvas: Canvas? = null

            // try locking the canvas for pixel editing
            try {
                canvas = this.surfaceHolder.lockCanvas()
                synchronized(surfaceHolder) {
                    this.gamePanel.update()
                    this.gamePanel.draw(canvas)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                if (canvas != null)
                    try {
                        surfaceHolder.unlockCanvasAndPost(canvas)
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
            }

            timeMillis = (System.nanoTime() - startTime) / 1000000
            waitTime = targetTime - timeMillis

            if (waitTime >= 0) {
                try {
                    sleep(waitTime)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
            //totalTime += System.nanoTime() - startTime;
            //frameCount++;
            //if (frameCount == FPS) {
            //    averageFPS = 1000/((totalTime/frameCount)/1000000);
            //    totalTime = 0;
            //    frameCount = 0;
            //    System.out.println("Average FPS: " + averageFPS);
            //}
        }
    }

    internal fun setRunning(run: Boolean) {
        running = run
    }

    companion object {
        private const val FPS = 30
        private var running: Boolean = false
    }
}
